lib/db2i.js

```diff
@@ -95,283 +94,149 @@ DB2.prototype.create = function(model, data, options, callback) {
+                                        ' FROM ' + tableName);
 
-  function executeWithConnection(connection, cb) {
-    // Execution for update requires running two
-    // separate SQL statements.  The second depends on the
-    // result of the first.
-    var countStmt = new ParameterizedSQL('SELECT COUNT(*) AS CNT FROM ');
-    countStmt.merge(tableName);
-    countStmt.merge(self.buildWhere(model, where));
-    countStmt.noResults = false;
+  selectStmt.merge(self.buildWhere(model, where));
+  selectStmt.merge(' FOR UPDATE');
+  self.parameterize(selectStmt);
 
-    connection.query(countStmt, function(err, countData) {
-      debug('DB2.prototype.replaceById (data): sql %j, countData=%j\n',
-            countStmt, countData);
+  var updateData;
 
+  var executeTransaction = function(connection, cb) {
+    connection.query(selectStmt.sql, selectStmt.params, function(err, data) {
+      debug('DB2iSeries.prototype.update stmt: %j data: %j', selectStmt, data);
       if (err) {
-        debug('DB2.prototype.replaceById (ERR): %j', err);
         return cb(err);
       }
 
-      if (countData[0]['CNT'] > 0) {
-        stmt = self.buildReplace(model, where, data, options);
-      } else {
-        data.count = 0;
-        return cb(err, data);
-      }
-
-      stmt.noResults = true;
-
-      connection.query(stmt, function(err, sData) {
-        if (err) {
-          debug('DB2.prototype.replaceById (ERR): %j', err);
-          return cb(err);
-        }
+      connection.query(updateStmt.sql, updateStmt.params,
+        function(err, updateData) {
+          debug('DB2iSeries.prototype.update stmt: %j data: %j', updateStmt,
+                updateData);
+          if (err) {
+            return cb(err);
+          }
 
-        data.count = countData[0]['CNT'];
-        cb(null, data);
-      });
+          return cb(null, data.length);
+        });
     });
   };
 
+  // If a transaction hasn't already been started, then start a local one now.
+  // We will have to deal with cleaning this up in the event some error
+  // occurs in the code below.
   if (options.transaction) {
-    executeWithConnection(options.transaction.connection,
-      function(err, data, meta) {
-        if (err) {
-          return callback && callback(err);
-        } else {
-          return callback && callback(null, data, meta);
-        }
-      });
+    executeTransaction(options.transaction.connection, function(err, retVal) {
+      if (err) {
+        return callback(err);
+      }
+      callback(null, {count: retVal});
+    });
   } else {
-    self.beginTransaction(Transaction.READ_COMMITTED, function(err, conn) {
+    self.beginTransaction(Transaction.REPEATABLE_READ, function(err, conn) {
       if (err) {
-        return callback && callback(err);
+        return callback(err);
+      } else {
+        executeTransaction(conn, function(err, retVal) {
+          if (err) {
+            self.rollback(conn, function(err) {});
```

Comment:
I don't like the code order here though and am changing to have the conn.close inside the rollback callback.